### DISCUSSION TITLE
pDebug->ReportLiveObjects(DXGI_DEBUG_ALL, DXGI_DEBUG_RLO_SUMMARY) throws exception in DEBUG mode when implementing Image Loading as detailed in https://github.com/ocornut/imgui/wiki/Image-Loading-and-Displaying-Examples#Example-for-DirectX12-users

### DISCUSSION THREAD

**Discussion by CalamityBarefoot** [ocornut/imgui#6204] [New Users Build/Link/Run issues ONLY!]
Hello,
I used the link below to load my own images file into a GPU texture for custom ImageButton.
It works like a charm!! Thanks to the author.
https://github.com/ocornut/imgui/wiki/Image-Loading-and-Displaying-Examples#Example-for-DirectX12-users
However, when I run in debug mode (only, release no issues), an exception is thrown when I close the main window (exit my ImGui app):
![image](https://user-images.githubusercontent.com/126497312/221614399-eb58feba-d646-4dca-9ecc-78fdc4505538.png)
To solve it temporarily I've commented off  pDebug->ReportLiveObjects(DXGI_DEBUG_ALL, DXGI_DEBUG_RLO_SUMMARY);
However, as I'm a complete noob, I'm worried that the excpetion may be important.
Is there anyone who understands where the excpetion comes from and what can be done to solve the issue?
Thanks in advance for any help.

**Comment by PathogenDavid** [comment]
> I'm worried that the excpetion may be important.

Your intuition is correct!

Exception `0x87A` is used by the Direct3D debug layer when break mode is enabled on a message. Chances are you (perhaps unintentionally) enabled break mode on a message emitted by `ReportLiveObjects`.

The break mode of a message is controlled using either `ID3D12InfoQueue`, `IDXGIInfoQueue`, or the extremely old and deprecated DirectX Control Panel utility (you probably don't even have that last one installed -- it's not even meant to be used with D3D12 although I think it still affects DXGI.)

Chances are you copy+pasted the `ID3D12InfoQueue` code from `example_win32_directx12`, which enables break on all warnings.

You can find the messages written by `ReportLiveObjects` in Visual Studio's debug output window. (If the pane is missing, go to `View` > `Output` and ensure `Show output from` is set to `Debug`. That's where the debug layer prints its messages.)

You will probably find something along the lines of:

```
D3D12 WARNING: Live ID3D12Device at 0x000001C6491CABA8, Refcount: 27 [ STATE_CREATION WARNING #274: LIVE_DEVICE]
D3D12: **BREAK** enabled for the previous message, which was: [ WARNING STATE_CREATION #274: LIVE_DEVICE ]
```

The proper fix is to find the object being leaked and make sure it gets cleaned up before your app exits. (If you name your objects using [`ID3D12Object::SetName`](https://learn.microsoft.com/en-us/windows/win32/api/d3d12/nf-d3d12-id3d12object-setname) that message will include that name, which helps a lot. Since you're following the texture tutorial you probably need to release `pTexture`.)

A not-so-proper non-fix is to disable breaking on warnings, but since you're still learning it's probably better to leave them on. (Once you get in the habit of paying attention to the debug output window it's not unreasonable to disable break for the warning severity, but you should definitely always fix those warnings eventually.)

----------

As an aside, Direct3D 12 is not very friendly towards beginners. You might consider using Direct3D 11 while you're still learning as the API is much more forgiving. (Just compare the Direct3D 11 texture upload example with the Direct3D 12 one.)

(The issue you ran into would've still happened with Direct3D 11, I'm just speaking generally.)

**Comment by CalamityBarefoot** [comment]
Thanks a million!!
The issue is fixed, I just needed to Release ID3D12Resource* my_texture01:
At the end of int main:
    // Cleanup
    my_texture01->Release();
    ImGui_ImplDX12_Shutdown();
    ImGui_ImplWin32_Shutdown();
    ImPlot::DestroyContext();
    ImGui::DestroyContext();
